### PR TITLE
Restrict access using Github collaborators

### PR DESCRIPTION
**Pull request by jsclayton**
## Description

This adds options that allows collaborators on a Github repository to be used to restrict access.

## Motivation and Context

We'd like to be able to use a **public** repository to allow access through the proxy since you can add outside collaborators to a public repository and not have them cost anything to your organization. If you add a collaborator to an org, team, or private repository that counts as a user seat and you are charged for it.

This also satisfies the feature request in #129. 

## How Has This Been Tested?

Using only the `-github-repo` flag, we use the access token of the authenticated user to check what access, if any, they have to the target repo. Since anyone can implicitly pull from a public repo, they must either have write access to a public repo, or any access to a private repo.

Being able to check for read only access of a public repo requires an extra flag, `-github-token`, which is a personal access token for a user that has write access to the repository. That token needs to be created with the `public_repo` scope. 

In addition to testing the various scenarios, I've added unit tests for the new functionality. It did require refactoring the http tests in the Github provider, but all tests continue to pass after the change.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by steakunderscore** [comment]
@jsclayton Mind merging master in to your branch to resolve the conflicts?

**Comment by jsclayton** [comment]
@steakunderscore rebased and fixed up. Not clear what to do about the code climate warning short of refactoring the Github API request bits for the whole file into a helper, which seems like it should be its own PR. 🤷

File: docs/2_auth.md [hunk 1]
```diff
@@ -101,11 +101,24 @@ Note: When using the Azure Auth provider with nginx and the cookie session store
...
 2.  Under `Authorization callback URL` enter the correct url ie `https://internal.yourcompany.com/oauth2/callback`
 
-The GitHub auth provider supports two additional parameters to restrict authentication to Organization or Team level access. Restricting by org and team is normally accompanied with `--email-domain=*`
+The GitHub auth provider supports two additional ways to restrict authentication to either organization and optional team level access, or to collaborators of a repository. Restricting by these options is normally accompanied with `--email-domain=*`
+
+To restrict by organization only, include the following flag:
 
     -github-org="": restrict logins to members of this organisation
+
+To restrict within an organization to specific teams, include the following flag in addition to `-github-org`:
```
Comment (steakunderscore) [inline docs/2_auth.md:110]: FYI we are only supporting flags with double dashes (`--`) as of a few days ago. But it looks like I missed a few in this file when I updated the docs. We should put up a septate PR to fix this.

**Comment by steakunderscore** [comment]
> Not clear what to do about the code climate warning short of refactoring the Github API request bits for the whole file into a helper, which seems like it should be its own PR. 
I agree, this should be its own PR. We only just turned on code climate, so it's picking up a few thing that need refactoring. Most of which we already knew about to be fair.

**Pull request merged**